crates/net/downloaders/src/test_utils/file_client.rs

```diff
@@ -62,19 +69,82 @@ impl FileClient {
     /// Create a new file client from a file path.
     pub async fn new<P: AsRef<Path>>(path: P) -> Result<Self, FileClientError> {
         let file = File::open(path).await?;
-        FileClient::from_file(file)
+        FileClient::from_file(file).await
     }
 
     /// Initialize the [`FileClient`](FileClient) with a file directly.
-    pub(crate) fn from_file(file: File) -> Result<Self, FileClientError> {
-        let reader = BufReader::new(file);
-
-        Ok(Self {
-            reader,
-            headers: HashMap::new(),
-            bodies: HashMap::new(),
-            is_syncing: Arc::new(Default::default()),
-        })
+    pub(crate) async fn from_file(file: File) -> Result<Self, FileClientError> {
+        // get file len from metadata before reading
+        let metadata = file.metadata().await?;
+        let file_len = metadata.len();
+
+        let mut reader = BufReader::new(file);
+
+        let mut headers = HashMap::new();
+        let mut hash_to_number = HashMap::new();
+        let mut bodies = HashMap::new();
+
+        // TODO: replace with builtin const fn for rlp headers?
+        // max rlp header length is 9
+        let max_header_len = 9;
+
+        let mut block_num = 0;
+        let mut cursor = 0u64;
+        while cursor < file_len {
+            // need to avoid eof when reading
+            let header_buf_len = std::cmp::min(max_header_len, file_len - cursor);
+            let mut block_buf = vec![0u8; header_buf_len as usize];
+            reader.read_exact(&mut block_buf).await?;
+
+            let block_buf_copy = block_buf.clone();
+
+            // TODO: change rlp header decode? could just read the entire thing into memory and not
+            // use a buffered reader...
+            //
+            // # SAFETY
+            // We know that we are reading a rlp header, but the rlp decoding will check the length
+            // of the buffer after decoding the payload length, to make sure the buffer is large
+            // enough to contain the object being decoded.
+            //
+            // We do not require this check, as it is done again when we decode the RawBlockBody.
+            // This is done to ensure that we can get the exact payload length of the block, so we
+            // can read the correct number of bytes from the file.
+            unsafe { block_buf.set_len(usize::MAX) };
+
```

Comment:
this seems dangerous, 
I don't understand why do we need to set the length manually?

Reply:
`RlpHeader::decode` will check the length of the buffer it's reading from and throw `InputTooShort` if it's not large enough. In this case we don't know how large the block is, so we don't know how many bytes to read. Decoding the rlp header makes sense for this, because it contains the payload length. 

Because of the length check, it will throw `InputTooShort` unless we change the length of the buffer or extend it to fit the whole block (but we don't know how large the block is until after we decode the rlp header)

Reply:
I'm even more confused now than before :D

Probably because I don't know the layout of the file content, but this seems way too complicated and wasteful.

so I should have asked, what's actually inside the file?

Reply:
The file is just rlp block bodies, one after the other

Reply:
a `RawBlockBody ` is encoded as list, right?

so we could just write a codec for this? and since file is asyncread, I guess we Could use `Framed` instead?

Reply:
> a RawBlockBody  is encoded as list, right?

yep, a RLP list

> so we could just write a codec for this? and since file is asyncread, I guess we Could use Framed instead?

yeah, that would be much nicer than the current method

Reply:
hmm, getting this after implementing a codec:
```
thread 'test_utils::file_client::tests::test_download_headers_from_file' panicked at 'called `Result::unwrap()` on an `Err` value: Rlp(InputTooShort)', crates/net/downloaders/src/test_utils/file_client.rs:323:65
```
https://github.com/paradigmxyz/reth/actions/runs/4068801528/jobs/7007785075#step:8:239

Will have to do some debugging, because there's no manual decoding here, but we return `InputTooShort` in `Header::decode` a lot so I suspect it's getting returned there

Reply:
ah, found the issue:
```
buf remaining: 325, payload length: 1009
```
https://github.com/paradigmxyz/reth/actions/runs/4069261533/jobs/7008790655#step:8:208

So because we're using a `BufReader`, when the rlp header is being decoded, the buf being passed to `decode` is not large enough to contain the entire `RawBlockBody` / payload. This causes the check in `decode` to fail:
https://github.com/paradigmxyz/reth/blob/51ab6fde9f18f0734b4a457afd97e032fe7c6c0c/crates/common/rlp/src/decode.rs#L147-L149

We can probably do one of the following:
 - add a method to `reth_rlp::Header` that decode the header without performing this check
 - instead of using a `BufReader`, load the entire file into memory all at once
 
 wdyt?

Reply:
After checking the files included in hive, they're small enough to just load the entire file without a BufReader. I guess if we want to load huge chain files at a later point, we'll have to revisit